File: samples/syntax/sample.php
<?php
/**
 * PHP Syntax Highlighting Test
 * 
 * This file demonstrates various PHP syntax constructs.
 */

declare(strict_types=1);

namespace App\Syntax;

use App\Interfaces\Stringable;
use App\Traits\Loggable;
use DateTime;
use DateTimeImmutable;
use Exception;
use InvalidArgumentException;
use JsonSerializable;
use PDO;
use RuntimeException;

// Constants
const VERSION = '1.0.0';
const PI = 3.14159265359;
const DEBUG = true;

define('APP_NAME', 'Syntax Test');
define('MAX_SIZE', 1024);

// Enums (PHP 8.1+)
enum Status: string
{
    case Pending = 'pending';
    case Active = 'active';
    case Completed = 'completed';
    case Failed = 'failed';

    public function label(): string
    {
        return match($this) {
            self::Pending => 'Waiting',
            self::Active => 'In Progress',
            self::Completed => 'Done',
            self::Failed => 'Error',
        };
    }
}

enum Color: int
{
    case Red = 1;
    case Green = 2;
    case Blue = 3;
}

// Interface
interface UserInterface
{
    public function getId(): int;
    public function getName(): string;
    public function getEmail(): ?string;
}

// Trait
trait TimestampTrait
{
    protected ?DateTimeImmutable $createdAt = null;
    protected ?DateTimeImmutable $updatedAt = null;

    public function getCreatedAt(): ?DateTimeImmutable
    {
        return $this->createdAt;
    }

    public function setCreatedAt(DateTimeImmutable $createdAt): void
    {
        $this->createdAt = $createdAt;
    }

    public function touch(): void
    {
        $this->updatedAt = new DateTimeImmutable();
    }
}

// Abstract class
abstract class Entity
{
    protected int $id;

    abstract public function validate(): bool;

    public function getId(): int
    {
        return $this->id;
    }
}

// Class with attributes (PHP 8.0+)
#[Attribute]
class Route
{
    public function __construct(
        public string $path,
        public array $methods = ['GET'],
    ) {}
}

// Main class
class User extends Entity implements UserInterface, JsonSerializable
{
    use TimestampTrait;

    public const ROLE_USER = 'user';
    public const ROLE_ADMIN = 'admin';

    private static int $instanceCount = 0;

    // Constructor property promotion (PHP 8.0+)
    public function __construct(
        protected string $name,
        protected string $email,
        protected string $role = self::ROLE_USER,
        private ?string $password = null,
        public readonly bool $active = true,
    ) {
        $this->id = ++self::$instanceCount;
        $this->createdAt = new DateTimeImmutable();
    }

    // Getters
    public function getId(): int
    {
        return $this->id;
    }

    public function getName(): string
    {
        return $this->name;
    }

    public function getEmail(): ?string
    {
        return $this->email;
    }

    // Setters with fluent interface
    public function setName(string $name): self
    {
        $this->name = $name;
        $this->touch();
        return $this;
    }

    public function setEmail(string $email): self
    {
        if (!filter_var($email, FILTER_VALIDATE_EMAIL)) {
            throw new InvalidArgumentException("Invalid email: $email");
        }
        $this->email = $email;
        $this->touch();
        return $this;
    }

    // Method with union types (PHP 8.0+)
    public function setPassword(string|null $password): void
    {
        $this->password = $password ? password_hash($password, PASSWORD_DEFAULT) : null;
    }

    // Method with named arguments
    public function greet(string $greeting = 'Hello', bool $formal = false): string
    {
        $title = $formal ? 'Mr./Ms. ' : '';
        return "$greeting, {$title}{$this->name}!";
    }

    // Static method
    public static function getInstanceCount(): int
    {
        return self::$instanceCount;
    }

    // Validation
    public function validate(): bool
    {
        return !empty($this->name) && !empty($this->email);
    }

    // JsonSerializable implementation
    public function jsonSerialize(): array
    {
        return [
            'id' => $this->id,
            'name' => $this->name,
            'email' => $this->email,
            'role' => $this->role,
            'active' => $this->active,
            'created_at' => $this->createdAt?->format('c'),
        ];
    }

    // Magic methods
    public function __toString(): string
    {
        return "User#{$this->id}: {$this->name}";
    }

    public function __get(string $name): mixed
    {
        return $this->$name ?? null;
    }

    public function __isset(string $name): bool
    {
        return isset($this->$name);
    }
}

// Generic-like class using templates
/**
 * @template T
 */
class Container
{
    /** @var array<int, T> */
    private array $items = [];

    /**
     * @param T $item
     */
    public function add(mixed $item): void
    {
        $this->items[] = $item;
    }

    /**
     * @return T|null
     */
    public function get(int $index): mixed
    {
        return $this->items[$index] ?? null;
    }

    public function count(): int
    {
        return count($this->items);
    }

    /**
     * @return array<int, T>
     */
    public function all(): array
    {
        return $this->items;
    }
}

// Controller with attributes
#[Route('/users')]
class UserController
{
    public function __construct(
        private readonly UserRepository $repository,
    ) {}

    #[Route('/users', methods: ['GET'])]
    public function index(): array
    {
        return $this->repository->findAll();
    }

    #[Route('/users/{id}', methods: ['GET'])]
    public function show(int $id): ?User
    {
        return $this->repository->find($id);
    }

    #[Route('/users', methods: ['POST'])]
    public function store(array $data): User
    {
        $user = new User(
            name: $data['name'],
            email: $data['email'],
        );
        $this->repository->save($user);
        return $user;
    }
}

// Functions
function add(int $a, int $b): int
{
    return $a + $b;
}

function multiply(float ...$numbers): float
{
    return array_reduce($numbers, fn($carry, $n) => $carry * $n, 1.0);
}

// Arrow functions (PHP 7.4+)
$double = fn(int $x): int => $x * 2;
$isEven = fn(int $x): bool => $x % 2 === 0;

// Closures
$greet = function (string $name) use ($greeting): string {
    return "$greeting, $name!";
};

$multiplier = function (int $factor): callable {
    return fn(int $x) => $x * $factor;
};

// Array functions
$numbers = [1, 2, 3, 4, 5];
$doubled = array_map(fn($n) => $n * 2, $numbers);
$evens = array_filter($numbers, fn($n) => $n % 2 === 0);
$sum = array_reduce($numbers, fn($carry, $n) => $carry + $n, 0);

// Match expression (PHP 8.0+)
function httpStatusText(int $code): string
{
    return match($code) {
        200 => 'OK',
        201 => 'Created',
        400 => 'Bad Request',
        404 => 'Not Found',
        500, 502, 503 => 'Server Error',
        default => 'Unknown',
    };
}

// Null safe operator (PHP 8.0+)
function getUserCity(?User $user): ?string
{
    return $user?->getAddress()?->getCity();
}

// Named arguments (PHP 8.0+)
$user = new User(
    name: 'Alice',
    email: 'alice@example.com',
    role: User::ROLE_ADMIN,
    active: true,
);

// Heredoc and Nowdoc
$html = <<<HTML
<div class="user">
    <h1>{$user->getName()}</h1>
    <p>{$user->getEmail()}</p>
</div>
HTML;

$sql = <<<'SQL'
SELECT * FROM users
WHERE active = 1
ORDER BY created_at DESC
SQL;

// Exception handling
try {
    $user->setEmail('invalid-email');
} catch (InvalidArgumentException $e) {
    echo "Validation error: " . $e->getMessage();
} catch (Exception $e) {
    echo "Error: " . $e->getMessage();
} finally {
    echo "Cleanup complete";
}

// Generators
function range_generator(int $start, int $end): Generator
{
    for ($i = $start; $i <= $end; $i++) {
        yield $i;
    }
}

function fibonacci(): Generator
{
    $a = 0;
    $b = 1;
    while (true) {
        yield $a;
        [$a, $b] = [$b, $a + $b];
    }
}

// Stringable interface (PHP 8.0+)
class Message implements \Stringable
{
    public function __construct(private string $content) {}

    public function __toString(): string
    {
        return $this->content;
    }
}

// Entry point
if (php_sapi_name() === 'cli') {
    echo "Running in CLI mode\n";
    
    // Create user
    $user = new User(
        name: 'Test User',
        email: 'test@example.com',
    );
    
    echo $user->greet(greeting: 'Welcome', formal: true) . "\n";
    echo json_encode($user, JSON_PRETTY_PRINT) . "\n";
}
?>

File: samples/syntax/sample.py
#!/usr/bin/env python3
"""
Python Syntax Highlighting Test

This module demonstrates various Python syntax constructs.
"""

from __future__ import annotations

import asyncio
import json
import os
import re
import sys
from abc import ABC, abstractmethod
from collections import defaultdict, namedtuple
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum, auto
from functools import lru_cache, wraps
from pathlib import Path
from typing import (
    Any,
    Callable,
    Dict,
    Generic,
    Iterator,
    List,
    Literal,
    Optional,
    Protocol,
    Tuple,
    TypeVar,
    Union,
    overload,
)

# Type variables
T = TypeVar('T')
K = TypeVar('K')
V = TypeVar('V')

# Constants
PI: float = 3.14159265359
MAX_SIZE: int = 1_000_000
DEBUG: bool = True
GREETING: str = "Hello, World!"

# Different number formats
decimal = 1_000_000
hexadecimal = 0xDEADBEEF
octal = 0o755
binary = 0b10101010
scientific = 1.5e-10
complex_num = 3 + 4j

# Strings
single_quoted = 'Single quotes'
double_quoted = "Double quotes"
raw_string = r"Raw \n string"
bytes_literal = b"Bytes literal"
f_string = f"Interpolated: {PI:.2f}"
multiline = """
    Multi-line
    string
"""


# Enums
class Color(Enum):
    RED = auto()
    GREEN = auto()
    BLUE = auto()

    def __str__(self) -> str:
        return self.name.lower()


class Status(Enum):
    PENDING = "pending"
    ACTIVE = "active"
    COMPLETED = "completed"


# Named tuple
Point = namedtuple('Point', ['x', 'y'])


# Dataclass
@dataclass
class Person:
    name: str
    age: int
    email: Optional[str] = None
    tags: List[str] = field(default_factory=list)
    created_at: datetime = field(default_factory=datetime.now)

    def __post_init__(self) -> None:
        if self.age < 0:
            raise ValueError("Age cannot be negative")

    @property
    def is_adult(self) -> bool:
        return self.age >= 18

    def greet(self) -> str:
        return f"Hello, I'm {self.name}!"


# Protocol (structural typing)
class Comparable(Protocol):
    def __lt__(self, other: Any) -> bool: ...
    def __eq__(self, other: Any) -> bool: ...


# Abstract base class
class Shape(ABC):
    @abstractmethod
    def area(self) -> float:
        pass

    @abstractmethod
    def perimeter(self) -> float:
        pass

    def describe(self) -> str:
        return f"Area: {self.area():.2f}, Perimeter: {self.perimeter():.2f}"


class Rectangle(Shape):
    def __init__(self, width: float, height: float) -> None:
        self._width = width
        self._height = height

    @property
    def width(self) -> float:
        return self._width

    @width.setter
    def width(self, value: float) -> None:
        if value <= 0:
            raise ValueError("Width must be positive")
        self._width = value

    def area(self) -> float:
        return self._width * self._height

    def perimeter(self) -> float:
        return 2 * (self._width + self._height)


class Circle(Shape):
    def __init__(self, radius: float) -> None:
        self.radius = radius

    def area(self) -> float:
        return PI * self.radius ** 2

    def perimeter(self) -> float:
        return 2 * PI * self.radius


# Generic class
class Container(Generic[T]):
    def __init__(self) -> None:
        self._items: List[T] = []

    def add(self, item: T) -> None:
        self._items.append(item)

    def get(self, index: int) -> T:
        return self._items[index]

    def __iter__(self) -> Iterator[T]:
        return iter(self._items)

    def __len__(self) -> int:
        return len(self._items)


# Decorators
def timer(func: Callable[..., T]) -> Callable[..., T]:
    @wraps(func)
    def wrapper(*args: Any, **kwargs: Any) -> T:
        import time
        start = time.perf_counter()
        result = func(*args, **kwargs)
        end = time.perf_counter()
        print(f"{func.__name__} took {end - start:.4f}s")
        return result
    return wrapper


def retry(max_attempts: int = 3, delay: float = 1.0):
    def decorator(func: Callable[..., T]) -> Callable[..., T]:
        @wraps(func)
        def wrapper(*args: Any, **kwargs: Any) -> T:
            for attempt in range(max_attempts):
                try:
                    return func(*args, **kwargs)
                except Exception as e:
                    if attempt == max_attempts - 1:
                        raise
                    import time
                    time.sleep(delay)
            raise RuntimeError("Should not reach here")
        return wrapper
    return decorator


# Function overloads
@overload
def process(value: str) -> str: ...
@overload
def process(value: int) -> int: ...
@overload
def process(value: list) -> list: ...


def process(value: Union[str, int, list]) -> Union[str, int, list]:
    if isinstance(value, str):
        return value.upper()
    elif isinstance(value, int):
        return value * 2
    else:
        return list(reversed(value))


# Cached function
@lru_cache(maxsize=128)
def fibonacci(n: int) -> int:
    if n < 2:
        return n
    return fibonacci(n - 1) + fibonacci(n - 2)


# Generator function
def count_up_to(n: int) -> Iterator[int]:
    i = 0
    while i < n:
        yield i
        i += 1


# Async functions
async def fetch_data(url: str) -> dict:
    await asyncio.sleep(0.1)  # Simulate network delay
    return {"url": url, "data": "sample"}


async def main() -> None:
    tasks = [fetch_data(f"https://api.example.com/{i}") for i in range(5)]
    results = await asyncio.gather(*tasks)
    for result in results:
        print(result)


# Context manager
class FileManager:
    def __init__(self, filename: str, mode: str = 'r') -> None:
        self.filename = filename
        self.mode = mode
        self.file = None

    def __enter__(self):
        self.file = open(self.filename, self.mode)
        return self.file

    def __exit__(self, exc_type, exc_val, exc_tb):
        if self.file:
            self.file.close()
        return False


# Pattern matching (Python 3.10+)
def http_status(status: int) -> str:
    match status:
        case 200:
            return "OK"
        case 201:
            return "Created"
        case 400:
            return "Bad Request"
        case 404:
            return "Not Found"
        case 500 | 502 | 503:
            return "Server Error"
        case _:
            return "Unknown"


def process_command(command: dict) -> str:
    match command:
        case {"action": "create", "name": name}:
            return f"Creating {name}"
        case {"action": "delete", "id": id}:
            return f"Deleting {id}"
        case {"action": action, **rest}:
            return f"Unknown action: {action}"
        case _:
            return "Invalid command"


# List comprehensions
squares = [x ** 2 for x in range(10)]
evens = [x for x in range(20) if x % 2 == 0]
matrix = [[i * j for j in range(5)] for i in range(5)]

# Dict and set comprehensions
word_lengths = {word: len(word) for word in ["hello", "world"]}
unique_chars = {char for char in "hello world"}

# Generator expression
sum_of_squares = sum(x ** 2 for x in range(1000))


# Exception handling
class CustomError(Exception):
    def __init__(self, message: str, code: int) -> None:
        super().__init__(message)
        self.code = code


def risky_operation(value: int) -> int:
    try:
        if value < 0:
            raise ValueError("Negative value")
        if value == 0:
            raise CustomError("Zero is not allowed", code=400)
        return 100 // value
    except ValueError as e:
        print(f"Value error: {e}")
        raise
    except CustomError as e:
        print(f"Custom error (code {e.code}): {e}")
        raise
    except ZeroDivisionError:
        print("Division by zero")
        return 0
    finally:
        print("Cleanup")


# Lambda functions
add = lambda x, y: x + y
double = lambda x: x * 2
is_even = lambda x: x % 2 == 0

# Higher-order functions
numbers = [1, 2, 3, 4, 5]
doubled = list(map(lambda x: x * 2, numbers))
filtered = list(filter(lambda x: x > 2, numbers))
from functools import reduce
total = reduce(lambda a, b: a + b, numbers, 0)


# Main entry point
if __name__ == "__main__":
    # Create instances
    person = Person("Alice", 30, "alice@example.com")
    rect = Rectangle(10, 5)
    circle = Circle(7)

    # Use pattern matching
    status = http_status(200)
    print(f"Status: {status}")

    # Run async
    asyncio.run(main())
